File: sdks/java/src/main/java/com/blok/nanoservice/server/RuntimeServer.java
package com.blok.nanoservice.server;

import com.blok.nanoservice.config.ServerConfig;
import com.blok.nanoservice.node.NodeRegistry;
import com.blok.nanoservice.types.ExecutionRequest;
import com.blok.nanoservice.types.ExecutionResult;
import com.blok.nanoservice.types.HealthStatus;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonSyntaxException;
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpServer;

import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.util.concurrent.Executors;

/**
 * HTTP server for the Blok nanoservice runtime.
 * Uses the JDK built-in {@code com.sun.net.httpserver.HttpServer}.
 * <p>
 * Exposes two endpoints:
 * <ul>
 *   <li>POST /execute - Execute a workflow node</li>
 *   <li>GET /health - Health check</li>
 * </ul>
 */
public class RuntimeServer {

    private final NodeRegistry registry;
    private final ServerConfig config;
    private final Gson gson;
    private HttpServer httpServer;

    /**
     * Creates a new RuntimeServer.
     *
     * @param registry the node registry
     * @param config   the server configuration
     * @throws IllegalArgumentException if registry or config is null
     */
    public RuntimeServer(NodeRegistry registry, ServerConfig config) {
        if (registry == null) {
            throw new IllegalArgumentException("registry must not be null");
        }
        if (config == null) {
            throw new IllegalArgumentException("config must not be null");
        }
        this.registry = registry;
        this.config = config;
        this.gson = new GsonBuilder()
                .serializeNulls()
                .create();
    }

    /**
     * Starts the HTTP server. Blocks until the server is stopped.
     *
     * @throws IOException if the server cannot bind to the port
     */
    public void start() throws IOException {
        InetSocketAddress address = new InetSocketAddress(config.getHost(), config.getPort());
        httpServer = HttpServer.create(address, 0);
        httpServer.setExecutor(Executors.newCachedThreadPool());

        httpServer.createContext("/execute", this::handleExecute);
        httpServer.createContext("/health", this::handleHealth);

        httpServer.start();

        System.out.println("Nanoservice runtime v" + config.getVersion()
                + " starting on " + config.address());
        System.out.println("Registered nodes: " + registry.nodeNames());
    }

    /**
     * Stops the HTTP server gracefully.
     */
    public void stop() {
        if (httpServer != null) {
            System.out.println("Shutting down nanoservice runtime...");
            httpServer.stop(config.getShutdownTimeoutSec());
        }
    }

    /**
     * Returns the underlying HttpServer for testing or advanced use.
     *
     * @return the HttpServer, or null if not started
     */
    public HttpServer getHttpServer() {
        return httpServer;
    }

    private void handleExecute(HttpExchange exchange) throws IOException {
        // CORS preflight
        if (config.isEnableCors()) {
            setCorsHeaders(exchange);
            if ("OPTIONS".equalsIgnoreCase(exchange.getRequestMethod())) {
                exchange.sendResponseHeaders(204, -1);
                exchange.close();
                return;
            }
        }

        if (!"POST".equalsIgnoreCase(exchange.getRequestMethod())) {
            ExecutionResult error = ExecutionResult.error("method not allowed, use POST");
            writeJson(exchange, 405, error);
            return;
        }

        try {
            InputStreamReader reader = new InputStreamReader(
                    exchange.getRequestBody(), StandardCharsets.UTF_8);
            ExecutionRequest request = gson.fromJson(reader, ExecutionRequest.class);

            if (request == null) {
                writeJson(exchange, 400, ExecutionResult.error("invalid JSON: empty body"));
                return;
            }

            ExecutionResult result = registry.execute(request);
            writeJson(exchange, 200, result);

        } catch (JsonSyntaxException e) {
            ExecutionResult error = ExecutionResult.error("invalid JSON: " + e.getMessage());
            writeJson(exchange, 400, error);
        } catch (Exception e) {
            ExecutionResult error = ExecutionResult.error("internal server error: " + e.getMessage());
            writeJson(exchange, 200, error);
        }
    }

    private void handleHealth(HttpExchange exchange) throws IOException {
        if (config.isEnableCors()) {
            setCorsHeaders(exchange);
            if ("OPTIONS".equalsIgnoreCase(exchange.getRequestMethod())) {
                exchange.sendResponseHeaders(204, -1);
                exchange.close();
                return;
            }
        }

        if (!"GET".equalsIgnoreCase(exchange.getRequestMethod())) {
            exchange.sendResponseHeaders(405, -1);
            exchange.close();
            return;
        }

        HealthStatus health = registry.health(config.getVersion());
        writeJson(exchange, 200, health);
    }

    private void writeJson(HttpExchange exchange, int statusCode, Object data) throws IOException {
        // Build JSON with null exclusion for optional fields
        Gson responseGson = new GsonBuilder().create();
        String json = responseGson.toJson(data);
        byte[] bytes = json.getBytes(StandardCharsets.UTF_8);

        exchange.getResponseHeaders().set("Content-Type", "application/json");
        exchange.sendResponseHeaders(statusCode, bytes.length);

        try (OutputStream os = exchange.getResponseBody()) {
            os.write(bytes);
        }
    }

    private void setCorsHeaders(HttpExchange exchange) {
        exchange.getResponseHeaders().set("Access-Control-Allow-Origin", "*");
        exchange.getResponseHeaders().set("Access-Control-Allow-Methods", "GET, POST, OPTIONS");
        exchange.getResponseHeaders().set("Access-Control-Allow-Headers", "Content-Type, Authorization");
    }
}

File: sdks/python3/blok/server/runtime_server.py
from __future__ import annotations
import json
import logging
import signal
import sys
from http.server import HTTPServer, BaseHTTPRequestHandler
from typing import Optional

from blok.node.node_registry import NodeRegistry
from blok.config.server_config import ServerConfig
from blok.types.execution_request import ExecutionRequest
from blok.types.execution_result import ExecutionResult

logger = logging.getLogger("blok")


class _RequestHandler(BaseHTTPRequestHandler):
    """HTTP request handler for the blok runtime."""

    registry: NodeRegistry
    config: ServerConfig

    def do_POST(self) -> None:
        if self.path != "/execute":
            self._write_json(
                404,
                ExecutionResult.error_result(f"not found: {self.path}").to_dict(),
            )
            return

        content_length = int(self.headers.get("Content-Length", 0))
        body = self.rfile.read(content_length)

        try:
            parsed = json.loads(body)
        except (json.JSONDecodeError, ValueError) as e:
            self._write_json(
                400,
                ExecutionResult.error_result(f"invalid JSON: {e}").to_dict(),
            )
            return

        execution_request = ExecutionRequest.from_dict(parsed)
        result = self.registry.execute(execution_request)
        self._write_json(200, result.to_dict())

    def do_GET(self) -> None:
        if self.path != "/health":
            self._write_json(404, {"error": f"not found: {self.path}"})
            return

        health = self.registry.health()
        self._write_json(200, health.to_dict())

    def _write_json(self, status_code: int, data: dict) -> None:
        response_body = json.dumps(data).encode("utf-8")

        self.send_response(status_code)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(response_body)))

        if self.config.enable_cors:
            self.send_header("Access-Control-Allow-Origin", "*")
            self.send_header("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
            self.send_header("Access-Control-Allow-Headers", "Content-Type, Authorization")

        self.end_headers()
        self.wfile.write(response_body)

    def do_OPTIONS(self) -> None:
        """Handle CORS preflight requests."""
        self.send_response(204)
        if self.config.enable_cors:
            self.send_header("Access-Control-Allow-Origin", "*")
            self.send_header("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
            self.send_header("Access-Control-Allow-Headers", "Content-Type, Authorization")
        self.end_headers()

    def log_message(self, format: str, *args) -> None:
        """Override to use Python logging instead of stderr."""
        logger.debug("%s %s", self.address_string(), format % args)


class RuntimeServer:
    """Blok HTTP server that handles execute and health requests.

    This server exposes two endpoints matching the Blok SDK contract:

    - ``POST /execute`` - Execute a node with the provided ExecutionRequest JSON
    - ``GET /health``   - Return runtime health status

    Usage::

        registry = NodeRegistry()
        registry.register("my-node", MyNode())

        server = RuntimeServer(registry)
        server.start()  # blocks until shutdown
    """

    def __init__(
        self,
        registry: NodeRegistry,
        config: Optional[ServerConfig] = None,
    ):
        self.registry = registry
        self.config = config or ServerConfig.from_env()
        self._server: Optional[HTTPServer] = None

    def start(self) -> None:
        """Start the HTTP server. Blocks until shutdown."""
        handler_class = type(
            "Handler",
            (_RequestHandler,),
            {"registry": self.registry, "config": self.config},
        )

        self._server = HTTPServer(
            (self.config.host, self.config.port),
            handler_class,
        )
        self._server.timeout = self.config.read_timeout_sec

        # Set up graceful shutdown.
        # NOTE: We use sys.exit() instead of self._server.shutdown() because
        # shutdown() must be called from a different thread than serve_forever().
        # Calling it from a signal handler in the same thread causes a deadlock.
        def shutdown_handler(signum, frame):
            logger.info("Shutdown signal received, stopping server...")
            sys.exit(0)

        signal.signal(signal.SIGTERM, shutdown_handler)
        signal.signal(signal.SIGINT, shutdown_handler)

        logger.info(
            "Blok runtime v%s starting on %s",
            self.config.version,
            self.config.address,
        )
        logger.info("Registered nodes: %s", self.registry.node_names())

        try:
            self._server.serve_forever()
        except KeyboardInterrupt:
            pass
        finally:
            self._server.server_close()
            logger.info("Server stopped.")

    def shutdown(self) -> None:
        """Gracefully shut down the server."""
        if self._server:
            self._server.shutdown()
